### PR TITLE
XCOMMONS-2016: Add a new escape method that only escape the minimal amount of XML special characters

### PR DESCRIPTION
**Pull request by manuelleduc**
https://jira.xwiki.org/browse/XCOMMONS-2016

### CODE REVIEW

File: xwiki-commons-core/xwiki-commons-xml/src/main/java/org/xwiki/xml/XMLUtils.java [hunk 1]
```diff
@@ -225,6 +225,7 @@ public static String unescapeXMLComment(String content)
      * @param content the text to escape, may be {@code null}
      * @return a new escaped {@code String}, {@code null} if {@code null} input
      */
+    @Deprecated
```
Comment (vmassol) [inline xwiki-commons-core/xwiki-commons-xml/src/main/java/org/xwiki/xml/XMLUtils.java:228]: As I mentioned on some chat, I'm not sure we wish to deprecate this. Note that whenever a deprecation is done it also requires a `@deprecated` javadoc tag with explanation as to what should be used and in which version it's been deprecated.

Comment (tmortagne) [reply xwiki-commons-core/xwiki-commons-xml/src/main/java/org/xwiki/xml/XMLUtils.java:228]: There is no reason to deprecate this.

File: xwiki-commons-core/xwiki-commons-xml/src/main/java/org/xwiki/xml/XMLUtils.java [hunk 2]
```diff
@@ -276,15 +277,15 @@ public static String escapeAttributeValue(Object content)
     }
 
     /**
-     *  Escapes some XML special characters (&amp; and &lt;) in a <code>String</code> using numerical XML
-     *  entities, so that the resulting string can safely be used in HTML pages.
+     * Escapes XML special characters in a {@code String} using numerical XML entities, so that the resulting string
+     * can safely be used as an XML text value. Specifically, escapes &lt;, and &amp;.
```
Comment (vmassol) [inline xwiki-commons-core/xwiki-commons-xml/src/main/java/org/xwiki/xml/XMLUtils.java:281]: "as an XML text value." --> "as an XML element text value (e.g. {@code Jim & John} will be escaped and can thus be put inside an XML tag, such as the {@code p} tag, as in {@code <p>Jim &amp; John</p>}).".

Should use `{@code ...}` constructs for `&lt;` and `&amp;`.

File: xwiki-commons-core/xwiki-commons-xml/src/main/java/org/xwiki/xml/XMLUtils.java [hunk 3]
```diff
@@ -276,15 +277,15 @@ public static String escapeAttributeValue(Object content)
...
      *
      * @param content the text to escape, may be {@code null}
      * @return a new escaped {@code String}, {@code null} if {@code null} input
      * @since 12.8RC1
      */
     @Unstable
-    public static String minimalEscape(Object content)
+    public static String escapeElementText(Object content)
```
Comment (vmassol) [inline xwiki-commons-core/xwiki-commons-xml/src/main/java/org/xwiki/xml/XMLUtils.java:288]: I think the javadoc for the content param should also be improved and we should explain why we accept an Object and not a String and what we do to it (say that we convert it to a String).

File: xwiki-commons-core/xwiki-commons-xml/src/main/java/org/xwiki/xml/html/HTMLUtils.java [hunk 4]
```diff
@@ -289,4 +291,18 @@ public static void stripFirstElementInside(Document document, String parentTagNa
...
+    /**
+     * Escapes HTML special characters in a {@code String} using numerical HTML entities, so that the resulting string
+     * can safely be used as an HTML content. Specifically, escapes &lt;, and &amp;.
+     *
+     * @param content the text to escape, may be {@code null}
+     * @return a new escaped {@code String}, {@code null} if {@code null} input
+     * @since 12.8RC1
+     */
+    @Unstable
+    public static String escapeElementContent(Object content)
```
Comment (vmassol) [inline xwiki-commons-core/xwiki-commons-xml/src/main/java/org/xwiki/xml/html/HTMLUtils.java:304]: Is it called Elements in HTML? We need to check this. Could be Tags (https://www.w3schools.com/tags/tag_div.asp calls it tags at least).

Comment (tmortagne) [reply xwiki-commons-core/xwiki-commons-xml/src/main/java/org/xwiki/xml/html/HTMLUtils.java:304]: It's called an "element" in all the XML APIs I can think of at least.

Comment (vmassol) [reply xwiki-commons-core/xwiki-commons-xml/src/main/java/org/xwiki/xml/html/HTMLUtils.java:304]: > It's called an "element" in all the XML APIs I can think of at least.

Yes but we're in HTMLUtils here. And it has its own spec and terminology.

File: xwiki-commons-core/xwiki-commons-xml/src/main/java/org/xwiki/xml/XMLUtils.java (hunk 1 above)
Comment (vmassol) [reply xwiki-commons-core/xwiki-commons-xml/src/main/java/org/xwiki/xml/XMLUtils.java:228]: Well there's a reason :) It's very unprecise/misleading and leads to extra escapes that are not needed in some situations. I'm fine to keep it but the javadoc needs to be improved to explain that (the extra escapes) and we need to recommended using the other 2 methods whenever possible.

File: xwiki-commons-core/xwiki-commons-xml/src/main/java/org/xwiki/xml/html/HTMLUtils.java (hunk 4 above)
Comment (manuelleduc) [reply xwiki-commons-core/xwiki-commons-xml/src/main/java/org/xwiki/xml/html/HTMLUtils.java:304]: I read some more doc and `Content` is too broad as it applied both to children tags and texts. `escapeElementText` is probably more accurate (and aligns with the XML vocabulary).

File: xwiki-commons-core/xwiki-commons-xml/src/main/java/org/xwiki/xml/XMLUtils.java (hunk 3 above)
Comment (manuelleduc) [reply xwiki-commons-core/xwiki-commons-xml/src/main/java/org/xwiki/xml/XMLUtils.java:298]: Documentation of the content parameter updated here https://github.com/xwiki/xwiki-commons/pull/102/files#diff-c1eed56adbad8fd67a8e196907ce032fR292

File: xwiki-commons-core/xwiki-commons-xml/src/main/java/org/xwiki/xml/XMLUtils.java (hunk 1 above)
Comment (manuelleduc) [reply xwiki-commons-core/xwiki-commons-xml/src/main/java/org/xwiki/xml/XMLUtils.java]: Deprecated annotation removed and javadoc updated here https://github.com/xwiki/xwiki-commons/pull/102/files#diff-c1eed56adbad8fd67a8e196907ce032fR221
- Some more explanations of why { is escaped
- Explanations of why is is best to use `escapeAttributeValue` and `escapeElementText` instead of `escape`

File: xwiki-commons-core/xwiki-commons-xml/src/main/java/org/xwiki/xml/XMLUtils.java (hunk 2 above)
Comment (manuelleduc) [reply xwiki-commons-core/xwiki-commons-xml/src/main/java/org/xwiki/xml/XMLUtils.java]: Javadoc updated here https://github.com/xwiki/xwiki-commons/pull/102/files#diff-c1eed56adbad8fd67a8e196907ce032fR287

File: xwiki-commons-core/xwiki-commons-xml/src/main/java/org/xwiki/xml/html/HTMLUtils.java (hunk 4 above)
Comment (manuelleduc) [reply xwiki-commons-core/xwiki-commons-xml/src/main/java/org/xwiki/xml/html/HTMLUtils.java]: Method renamed to `escapeElementText` and javadoc updated here https://github.com/xwiki/xwiki-commons/pull/102/files#diff-6866bfd5f73fabc8380825bf60260423R307

File: xwiki-commons-core/xwiki-commons-xml/src/main/java/org/xwiki/xml/XMLUtils.java [hunk 5]
```diff
@@ -218,12 +218,20 @@ public static String unescapeXMLComment(String content)
     }
 
     /**
-     * Escapes all the XML special characters in a <code>String</code> using numerical XML entities. Specifically,
-     * escapes &lt;, &gt;, ", ', &amp; and {.  Left curly bracket is included here to protect against {{/html}} in
-     * xwiki 2.x syntax.
+     * <p>
+     * Escapes all the XML special characters and a xwiki 2.x special character (i.e., <code>{</code>) in a
```
Comment (vmassol) [inline xwiki-commons-core/xwiki-commons-xml/src/main/java/org/xwiki/xml/XMLUtils.java:222]: What is a "xwiki 2.x special character"? Maybe you mean "a XWiki Syntax 2.0+ special character"? Now this is already mentioned at the end of the sentence. 

Note: I think `</p>` leads to error in the javadoc generation but I'm not sure. I don't think tags must be closed in javadoc (it's not pure HTML). See for ex https://blog.joda.org/2012/11/javadoc-coding-standards.html

I think the javadoc still missing explanation as to why you would use this method. See my proposals on the chat.

File: xwiki-commons-core/xwiki-commons-xml/src/main/java/org/xwiki/xml/html/HTMLUtils.java [hunk 6]
```diff
@@ -289,4 +291,21 @@ public static void stripFirstElementInside(Document document, String parentTagNa
...
         }
     }
+
+    /**
+     * Escapes HTML special characters in a {@code String} using numerical HTML entities, so that the resulting string
+     * can safely be used as an HTML content text value. For instance, {@code Jim & John} will be escaped and can thus
+     * be put inside a HTML tag, such as the {@code p} tag, as in {@code <p>Jim &amp; John</p>}. Specifically, escapes
+     * &lt; to {@code &lt;}, and &amp; to {@code &amp;}.
+     *
+     * @param content the text to escape, may be {@code null}. The content is converted to a {@link String} using
```
Comment (vmassol) [inline xwiki-commons-core/xwiki-commons-xml/src/main/java/org/xwiki/xml/html/HTMLUtils.java:301]: Still missing the why? Why did you add a signature that takes an Object instead of a String!

Comment (manuelleduc) [reply xwiki-commons-core/xwiki-commons-xml/src/main/java/org/xwiki/xml/html/HTMLUtils.java:301]: I made `content` an `Object` to stay consistent with the other escape methods. But AFAIK we would be fine with a `content` of type `String`. Would it be ok to change the signature of `escapeElementText` to `public static String escapeElementText(String content)`?

Comment (vmassol) [reply xwiki-commons-core/xwiki-commons-xml/src/main/java/org/xwiki/xml/html/HTMLUtils.java:301]: > I made content an Object to stay consistent with the other escape methods.

But that's not a good-enough reason :) Why did they do that? Imagine they made a mistake, you'd just be continuing the mistake.

Comment (vmassol) [reply xwiki-commons-core/xwiki-commons-xml/src/main/java/org/xwiki/xml/html/HTMLUtils.java:301]: I think it comes from Velocity.

See:
* Taken from https://github.com/xwiki/xwiki-platform/commit/ed3a5467b08c7069959076ace8e40a2c17bb253c#diff-83c0483287bd455e80eb573959efc65bR47
* And move to https://github.com/xwiki/xwiki-platform/commit/ed3a5467b08c7069959076ace8e40a2c17bb253c#diff-2c6a53f98aefb9159a01c42a82e54255R177

Now need to understand why Velocity tool is using Object. And whether it's needed for a Java API (I don't think so).

Comment (vmassol) [reply xwiki-commons-core/xwiki-commons-xml/src/main/java/org/xwiki/xml/html/HTMLUtils.java:301]: Not sure what to do. I think i'd be in favor of using String for the new method instead of Object. WDYT?

Comment (manuelleduc) [reply xwiki-commons-core/xwiki-commons-xml/src/main/java/org/xwiki/xml/html/HTMLUtils.java:301]: - We control the places where the method is called (and they are all passing string values)
- IMHO, it is a good practice to force the caller to sanitize its input before (instead of dealing with unexpected value in a method).

I'm in favor of using String for the new method.

Comment (vmassol) [reply xwiki-commons-core/xwiki-commons-xml/src/main/java/org/xwiki/xml/html/HTMLUtils.java:301]: Yep, let's see if @tmortagne and @surli are ok on this too.

Comment (vmassol) [reply xwiki-commons-core/xwiki-commons-xml/src/main/java/org/xwiki/xml/html/HTMLUtils.java:301]: Otherwise it would mean that everywhere we take a String in APIs we should move to Object and convert to a String. Feel not great and not typed ;)

Comment (vmassol) [reply xwiki-commons-core/xwiki-commons-xml/src/main/java/org/xwiki/xml/html/HTMLUtils.java:301]: We should probably introduce new signatures for existing methods taking an Object and deprecate existing ones (AFAIK specific types are favored over Object in Java so it should work).

Comment (tmortagne) [reply xwiki-commons-core/xwiki-commons-xml/src/main/java/org/xwiki/xml/html/HTMLUtils.java:301]: I can understand it for Velocity but I don't like this Object parameter for a Java utility method.

File: xwiki-commons-core/xwiki-commons-xml/src/main/java/org/xwiki/xml/XMLUtils.java (hunk 5 above)
Comment (manuelleduc) [reply xwiki-commons-core/xwiki-commons-xml/src/main/java/org/xwiki/xml/XMLUtils.java]: Javadoc update to use better writing standard + adding more explanation on why each method should be used for

File: xwiki-commons-core/xwiki-commons-xml/src/main/java/org/xwiki/xml/html/HTMLUtils.java (hunk 6 above)
Comment (manuelleduc) [reply xwiki-commons-core/xwiki-commons-xml/src/main/java/org/xwiki/xml/html/HTMLUtils.java]: Each method taking an Object in parameter is now deprecated and is overridden with a method taking a String in parameter

File: xwiki-commons-core/xwiki-commons-xml/src/main/java/org/xwiki/xml/XMLUtils.java [hunk 7]
```diff
@@ -349,24 +448,42 @@ public static String escapeElementContent(Object content)
...
      * ones that affect the resulting markup.
      *
-     * @param content the text to decode, may be {@code null}
+     * @param content the text to decode, may be {@code null}. The content is converted to {@code String} using
+     * {@link String#valueOf(Object)} before escaping.
      * @return unescaped content, {@code null} if {@code null} input
+     * @deprecated since 12.8RC1, use {@link #unescape(String)} instead
      */
+    @Deprecated
     public static String unescape(Object content)
```
Comment (vmassol) [inline xwiki-commons-core/xwiki-commons-xml/src/main/java/org/xwiki/xml/XMLUtils.java:457]: I think it should be easy in your IDE to find all existing calls to `unescape(Object)` in our code base and replace them with calls to `unescape(String)`. This would allow you/us to move the deprecated method to legacy (see "Second Step" in https://dev.xwiki.org/xwiki/bin/view/Community/DevelopmentPractices#HBackwardCompatibility).

WDYT?

Comment (manuelleduc) [reply xwiki-commons-core/xwiki-commons-xml/src/main/java/org/xwiki/xml/XMLUtils.java:457]: Doing it. 
Also:
1. I forgot to add javadoc `@since` for the new methods (fixing)
2. Should I add an `@Unstable` annotation on the new methods?

Comment (vmassol) [reply xwiki-commons-core/xwiki-commons-xml/src/main/java/org/xwiki/xml/XMLUtils.java:457]: > Should I add an @Unstable annotation on the new methods?

Yes

Comment (vmassol) [reply xwiki-commons-core/xwiki-commons-xml/src/main/java/org/xwiki/xml/XMLUtils.java:457]: > Doing it.

Cool, thanks!

Comment (manuelleduc) [reply xwiki-commons-core/xwiki-commons-xml/src/main/java/org/xwiki/xml/XMLUtils.java:464]: Deprecated and since added.

Comment (manuelleduc) [reply xwiki-commons-core/xwiki-commons-xml/src/main/java/org/xwiki/xml/XMLUtils.java:464]: Call to the deprecated methods updated in https://github.com/xwiki/xwiki-platform/pull/1384

File: xwiki-commons-core/xwiki-commons-velocity/src/main/java/org/xwiki/velocity/tools/EscapeTool.java [hunk 8]
```diff
@@ -68,7 +69,7 @@
     @Override
     public String xml(Object content)
     {
-        return XMLUtils.escape(content);
+        return XMLUtils.escape(Objects.toString(content, null));
```
Comment (surli) [inline xwiki-commons-core/xwiki-commons-velocity/src/main/java/org/xwiki/velocity/tools/EscapeTool.java:72]: Second argument is completely useless there no? According to javadoc it's exactly what does `Objects.toString` with only one argument.

File: xwiki-commons-core/xwiki-commons-xml/src/main/java/org/xwiki/xml/XMLUtils.java [hunk 9]
```diff
@@ -218,37 +219,121 @@ public static String unescapeXMLComment(String content)
...
+     * /&gt;</code>.
+     * <p>
+     * Specifically, escapes &lt;, &gt;, ", ', &amp; and {.
+     * <p>
+     * Note that is is preferable to use {@link #escapeAttributeValue(String)} when the content is used as
+     * an XML tag attribute, and {@link #escapeElementText(String)} when the content is used as an XML text.
      *
-     * @param content the text to escape, may be {@code null}
+     * @param content the text to escape, may be {@code null}. The content is converted to {@code String} using
+     * {@link String#valueOf(Object)} before escaping.
```
Comment (surli) [inline xwiki-commons-core/xwiki-commons-xml/src/main/java/org/xwiki/xml/XMLUtils.java:241]: You're actually using `Objects.toString()` here.

**Review by surli — APPROVED** [review]

File: xwiki-commons-core/xwiki-commons-xml/src/main/java/org/xwiki/xml/XMLUtils.java [hunk 10]
```diff
@@ -314,20 +401,39 @@ public static String minimalEscape(Object content)
...
+    /**
+     * Escapes the XML special characters in a <code>String</code> using numerical XML entities, so that the resulting
+     * string can safely be used as an XML text node. Specifically, escapes &lt;, &gt;, and &amp;.
+     *
+     * @param content the text to escape, may be {@code null}
+     * @return a new escaped {@code String}, {@code null} if {@code null} input
+     * @since 12.8RC1
+     */
+    @Unstable
+    public static String escapeElementContent(String content)
```
Comment (mflorea) [inline xwiki-commons-core/xwiki-commons-xml/src/main/java/org/xwiki/xml/XMLUtils.java:427]: It's not clear at all from the JavaDoc **why** we need both ``escapeElementText`` and ``escapeElementContent``. The JavaDoc explains what each method does and I understand that the only difference is that ``escapeElementText`` doesn't escape ``>`` (``&gt;``) but when would you use one versus the other? When is it OK to not escape ``>`` and when is it mandatory to escape it?

File: xwiki-commons-core/xwiki-commons-velocity/src/main/java/org/xwiki/velocity/tools/EscapeTool.java (hunk 8 above)
Comment (manuelleduc) [reply xwiki-commons-core/xwiki-commons-velocity/src/main/java/org/xwiki/velocity/tools/EscapeTool.java:72]:
```
jshell> Objects.toString(null);
$1 ==> "null"

jshell> Objects.toString(null, null);
$2 ==> null
```

`Objects#toString(Object)` returns the `"null"` string in case of a null parameter.

File: xwiki-commons-core/xwiki-commons-xml/src/main/java/org/xwiki/xml/XMLUtils.java (hunk 10 above)
Comment (tmortagne) [reply xwiki-commons-core/xwiki-commons-xml/src/main/java/org/xwiki/xml/XMLUtils.java:427]: Actually escaping `>` is never really needed so I don't understand why we need escapeElementContent at all.

Comment (manuelleduc) [reply xwiki-commons-core/xwiki-commons-xml/src/main/java/org/xwiki/xml/XMLUtils.java:427]: Would it be ok to deprecate `escapeElementContent` and suggest to use `escapeElementText` instead?

Comment (tmortagne) [reply xwiki-commons-core/xwiki-commons-xml/src/main/java/org/xwiki/xml/XMLUtils.java:427]: Sounds good.

Comment (manuelleduc) [reply xwiki-commons-core/xwiki-commons-xml/src/main/java/org/xwiki/xml/XMLUtils.java:427]: Hum, in this case we are in a odd situation when `public static String escapeElementContent(String content)` will be both `@Unstable` and `@Deprecated`. Does deprecated takes precedence on unstable?

Comment (tmortagne) [reply xwiki-commons-core/xwiki-commons-xml/src/main/java/org/xwiki/xml/XMLUtils.java:427]: I don't understand what you mean, `escapeElementContent(String content)` should not exist at all.

Comment (manuelleduc) [reply xwiki-commons-core/xwiki-commons-xml/src/main/java/org/xwiki/xml/XMLUtils.java]: `escapeElementContent` deprecated and `escapeElementText` recommended instead

File: xwiki-commons-core/xwiki-commons-xml/src/main/java/org/xwiki/xml/XMLUtils.java [hunk 11]
```diff
@@ -218,37 +219,121 @@ public static String unescapeXMLComment(String content)
...
      * @return a new escaped {@code String}, {@code null} if {@code null} input
+     * @see #escapeAttributeValue(String)
+     * @see #escapeElementText(String)
+     * @deprecated since 12.8RC1, use {@link #escape(String)} instead
      */
+    @Deprecated
     public static String escape(Object content)
     {
-        return escapeAttributeValue(content);
+        return escapeAttributeValue(Objects.toString(content, null));
```
Comment (mflorea) [inline xwiki-commons-core/xwiki-commons-xml/src/main/java/org/xwiki/xml/XMLUtils.java:250]: I would call ``escape`` here to be in line with the ``@deprecated`` message and to make sure that both ``escape`` methods behave the same (someone may update ``escape(String)`` without updating the deprecated ``escape(Object)``). This is minor but it's a best practice IMO.

File: xwiki-commons-core/xwiki-commons-xml/src/main/java/org/xwiki/xml/XMLUtils.java [hunk 12]
```diff
@@ -349,24 +439,44 @@ public static String escapeElementContent(Object content)
...
-        str = QUOT_PATTERN.matcher(str).replaceAll("\"");
-        str = LT_PATTERN.matcher(str).replaceAll("<");
-        str = GT_PATTERN.matcher(str).replaceAll(">");
-        str = AMP_PATTERN.matcher(str).replaceAll("&");
-        str = LCURL_PATTERN.matcher(str).replaceAll("{");
+        return unescape(String.valueOf(content));
+    }
+
+    /**
+     * Unescape encoded special XML characters. Only &gt;, &lt; &amp;, ", ' and { are unescaped, since they are the only
```
Comment (tmortagne) [inline xwiki-commons-core/xwiki-commons-xml/src/main/java/org/xwiki/xml/XMLUtils.java:458]: There is no reason to provide a custom unecaper, anyone with the need to unecape XML should use one of the many existing XML unescaper.

Comment (manuelleduc) [reply xwiki-commons-core/xwiki-commons-xml/src/main/java/org/xwiki/xml/XMLUtils.java]: Unescaper deprecated and `org.apache.commons.text.StringEscapeUtils#unescapeXml(String)` suggested insted.

File: xwiki-commons-core/xwiki-commons-xml/src/main/java/org/xwiki/xml/XMLUtils.java (hunk 11 above)
Comment (manuelleduc) [reply xwiki-commons-core/xwiki-commons-xml/src/main/java/org/xwiki/xml/XMLUtils.java]: I agree, `escape(Object)` is now calling `escape(String)` which itself calls `escapeAttributeValue(String)`.

**Review by tmortagne — APPROVED** [review]

**Review by mflorea — APPROVED** [review]
